File: cpp/cpp_100-999/890.cc
// C++
// leetcode 890
// https://leetcode.cn/problems/find-and-replace-pattern/

class Solution {
public:
    vector<string> findAndReplacePattern(vector<string>& words, string pattern) {
        int a[26], b[26], n = words.size(), m = pattern.size();
        memset(a, -1, sizeof(a));
        vector<string> res;
        for(int i = 0; i < n; ++i) {
            string &s = words[i];
            if(s.size() != m)
                continue;
            memset(a, -1, sizeof(a));
            memset(b, -1, sizeof(b));
            bool ok = true;
            for(int j = 0; j < m; j++) {
                int c1 = s[j] - 'a', c2 = pattern[j] - 'a';
                if(a[c1] == -1 && b[c2] == -1) {
                    a[c1] = c2;
                    b[c2] = c1;
                } else if(a[c1] == c2 && b[c2] == c1) {
                    // do nothing
                } else {
                    ok = false;
                    break;
                }
            }
            if(ok)
                res.push_back(s);
        }
        return res;
    }
};

File: Go/Go_100-999/890.go
// Go
// leetcode 890
// https://leetcode.cn/problems/find-and-replace-pattern/

func findAndReplacePattern(words []string, pattern string) []string {
	m :=  len(pattern)
	var a, b [26]int
	var res []string
	out:
		for _, s := range words {
			if len(s) != m {
				continue
			}
			for i := 0; i < 26; i++ {
				a[i] = -1
				b[i] = -1
			}
			for i := 0; i < m; i++ {
				u1, u2 := int(pattern[i] - 'a'), int(s[i] - 'a')
				if a[u1] == -1 && b[u2] == -1 {
					a[u1] = u2
					b[u2] = u1
				} else if a[u1] == u2 && b[u2] == u1 {
					continue
				} else {
					continue out
				}
			}
			res = append(res, s)
		}
	return res
}